### PR TITLE
KAFKA-2998: log warnings when client is disconnected from bootstrap brokers

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/Cluster.java [hunk 1]
```diff
@@ -28,13 +28,14 @@
...
+    private final boolean isBootstrapConfigured;
     private final List<Node> nodes;
     private final Set<String> unauthorizedTopics;
     private final Map<TopicPartition, PartitionInfo> partitionsByTopicPartition;
     private final Map<String, List<PartitionInfo>> partitionsByTopic;
     private final Map<String, List<PartitionInfo>> availablePartitionsByTopic;
-    private final Map<Integer, List<PartitionInfo>> partitionsByNode;
-    private final Map<Integer, Node> nodesById;
+    private final Map<String, List<PartitionInfo>> partitionsByNodeId;
+    private final Map<String, Node> nodesById;
```
Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: Was it required to change all the ids to strings?

**Comment by granthenke** [comment]
Is it relatively simple to add a test for the new behavior? I think this change might also solve [KAFKA-3468](https://issues.apache.org/jira/browse/KAFKA-3468).

File: clients/src/main/java/org/apache/kafka/common/Cluster.java (hunk 1 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: I needed to have `nodesById` use a String since I needed to be able to get the `Node` from a RequestSend object, which only has access to the string version. I changed the other one as well since it looked a little funny being the only one using an integer key. One nice refactor might be to merge these into a single map with a container object containing the different values.

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: Can we use "Integer.parseInt()" instead?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: That or maybe we could update RequestSend to store the node id?

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: It does seem like we could pass the node id to `RequestSend` without much issue.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: I'm actually tempted to call parseInt() within NetworkClient and revert the changes here in order to reduce the scope of this patch.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Cluster.java]: That's OK too.

**Comment by guozhangwang** [comment]
Filed https://issues.apache.org/jira/browse/KAFKA-3508 for unit test failure.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk.
